attrd/commands.c

```diff
@@ -718,6 +748,68 @@ attrd_lookup_or_create_value(GHashTable *values, const char *host, xmlNode *xml)
     return(v);
 }
 
+static gboolean
+check_current_attribute_from_sync_response(crm_node_t *peer, xmlNode *xml, const char *search_host, const char *name, const char *current_value)
+{
+    xmlNode *child = NULL;
+    gboolean find = FALSE;
+
+    for (child = __xml_first_child(xml); child != NULL; child = __xml_next(child)) {
+        const char *host = crm_element_value(child, F_ATTRD_HOST);
+        const char *attr = crm_element_value(child, F_ATTRD_ATTRIBUTE);
+        const char *value = crm_element_value(xml, F_ATTRD_VALUE);
+
+        if (safe_str_eq(host, search_host)) {
+            if (safe_str_eq(attr, name)) {
+                crm_debug("Sync response attribute(%s[%s] = %s) was ignore.(already stored) : [%s] = %s", attr, host, value, search_host, current_value);
+                find = TRUE;
+                break;
+            }
+        } else {
+            crm_debug("Sync response attribute(%s[%s] = %s) was ignore.(from another host) : [%s]", attr, host, value, search_host);
+        }
+    }
+    return find;
+}
+
+void 
+attrd_current_only_attribute_update(crm_node_t *peer, xmlNode *xml)
+{
+    GHashTableIter aIter;
+    GHashTableIter vIter;
+    attribute_t *a;
+    attribute_value_t *v = NULL;
+    xmlNode *sync = create_xml_node(NULL, __FUNCTION__);
+    gboolean build = FALSE;    
+
+    crm_xml_add(sync, F_ATTRD_TASK, ATTRD_OP_SYNC_RESPONSE);
+
+    g_hash_table_iter_init(&aIter, attributes);
+    while (g_hash_table_iter_next(&aIter, NULL, (gpointer *) & a)) {
+        g_hash_table_iter_init(&vIter, a->values);
+        while (g_hash_table_iter_next(&vIter, NULL, (gpointer *) & v)) {
+            if (safe_str_eq(v->nodename, attrd_cluster->uname) && v->seen == FALSE) {
+                if (check_current_attribute_from_sync_response(peer, xml, attrd_cluster->uname, a->id, v->current) == FALSE) {
```

Comment:
I believe check_current_attribute_from_sync_response() is now unnecessary. If v->seen is false, we know the XML doesn't contain the attribute.

Reply:
Hi Ken,

Thanks for your comment.
I think that the following condition judgment is unnecessary, but it seems to be a problem to not execute check_current_attribute_from_sync_response.

```
+            if (safe_str_eq(attr, name)) {
+                crm_debug("Sync response attribute(%s[%s] = %s) was ignore.(already stored) : [%s] = %s", attr, host, value, search_host, current_value);
+                find = TRUE;
+                break;
+            }
```

I will check it.

Best Regards,
Hideo Yamauchi.

Reply:
Hi Ken,

Your point is right.
The Check_current_attribute_from_sync_response processing is deprecated.

```
            if (safe_str_eq(v->nodename, attrd_cluster->uname) && v->seen == FALSE) {
                 crm_trace("Syncing %s[%s] = %s to everyone.(from local only attributes)", a->id, v->nodename, v->current);

```

I will change some crm_debugs to crm_trace.

How about with this change?

*Although the talk changes slightly, is it schedule for release of PM 1.1.19?
*We want this modification to be included in PM 1.1 series.

Best Regards,
Hideo Yamauchi.

Reply:
Hi Hideo,

Yes, that sounds good. It should be merged into 2.0 first, and then backported to the 1.1 branch. I have some other backporting I will need to do soon, so I can handle that part.

Reply:
Hi Ken,

All right!

Many thanks!
Hideo Yamauchi.